Script to dump content into an S3 bucket

mmulich opened the pull request:
See the readme for basic information. We can adjust it as needed.

mmulich committed:
Script to dump content into an S3 bucket

karenc commented on dump/README.md:
````diff
@@ -0,0 +1,45 @@
+```sh
+pip install -r requirements.txt
+```
+
````
May be we can add a link to boto3 quick start for the aws set up stuff https://github.com/boto/boto3#quick-start

karenc commented on dump/dump-to-bucket.py:
```diff
@@ -0,0 +1,296 @@
+    yield get_content(url, type_, format_, temperature, ident_hash_seq, 'application/json', -1)
+
+    # Save the raw json for later
+    raw_json = resp.json()
```
Mike got the same error:

```
Dumping raw-book-json into the bucket at "raw/02776133-d49d-49cb-bfaa-67c7f61b25a1@8.10.json" (application/json)
Traceback (most recent call last):
  File "./dump-to-bucket.py", line 296, in <module>
    main()
  File "/tmp/rap-spike-lambda/py3/lib/python3.7/site-packages/click/core.py", line 764, in __call__
    return self.main(*args, **kwargs)
  File "/tmp/rap-spike-lambda/py3/lib/python3.7/site-packages/click/core.py", line 717, in main
    rv = self.invoke(ctx)
  File "/tmp/rap-spike-lambda/py3/lib/python3.7/site-packages/click/core.py", line 956, in invoke
    return ctx.invoke(self.callback, **ctx.params)
  File "/tmp/rap-spike-lambda/py3/lib/python3.7/site-packages/click/core.py", line 555, in invoke
    return callback(*args, **kwargs)
  File "./dump-to-bucket.py", line 292, in main
    dump_in_bucket(scrape(book, host), (bucket, region))
  File "./dump-to-bucket.py", line 268, in dump_in_bucket
    for item in items:
  File "./dump-to-bucket.py", line 122, in scrape
    for item in scrape_version(id, version, host, visited_locs):
  File "./dump-to-bucket.py", line 216, in scrape_version
    raw_json = resp.json()
UnboundLocalError: local variable 'resp' referenced before assignment
```

At first I thought I can add `resp = yield get_content(...` on line 213 but that's not it...  Need to read more

karenc replied:
Ok I get it!  The `resp.json()` lines just need to be removed or commented out.

mmulich replied:
Yeah, I copied the wrong version here. I'll need to take out line 213 and uncomment the stuff above it. You can see this repeatedly in the next few statements too. I tried to refactor and forgot I needed access to the response.

karenc commented on dump/requirements.txt:
```diff
@@ -0,0 +1,4 @@
+blessings
+boto3
+click
+requests
```
Missing `cnx-common` in this list

mmulich replied:
O... my. The sloppiness of all this is terrible. 😝

karenc commented on dump/dump-to-bucket.py:
```diff
@@ -0,0 +1,296 @@
+
+    """
+    if not type_.startswith('resource'):
+        ids = [join_ident_hash(*i) for i in ident]
```
```
Traceback (most recent call last):
  File "./dump-to-bucket.py", line 290, in <module>
    main()
  File "/tmp/rap-spike-lambda/py3/lib/python3.7/site-packages/click/core.py", line 764, in __call__
    return self.main(*args, **kwargs)
  File "/tmp/rap-spike-lambda/py3/lib/python3.7/site-packages/click/core.py", line 717, in main
    rv = self.invoke(ctx)
  File "/tmp/rap-spike-lambda/py3/lib/python3.7/site-packages/click/core.py", line 956, in invoke
    return ctx.invoke(self.callback, **ctx.params)
  File "/tmp/rap-spike-lambda/py3/lib/python3.7/site-packages/click/core.py", line 555, in invoke
    return callback(*args, **kwargs)
  File "./dump-to-bucket.py", line 286, in main
    dump_in_bucket(scrape(book, host), (bucket, region))
  File "./dump-to-bucket.py", line 264, in dump_in_bucket
    key = gen_filepath(type, ident)
  File "./dump-to-bucket.py", line 62, in gen_filepath
    ids = [join_ident_hash(*i) for i in ident]
  File "./dump-to-bucket.py", line 62, in <listcomp>
    ids = [join_ident_hash(*i) for i in ident]
TypeError: join_ident_hash() missing 1 required positional argument: 'version'
```

I did a bit of debugging and `i` looks like this:

```
latest version of requested book: 02776133-d49d-49cb-bfaa-67c7f61b25a1@8.10
Requesting raw json book at https://archive.cnx.org/contents/02776133-d49d-49cb-bfaa-67c7f61b25a1@8.10.json?as_collated=0
i: ('02776133-d49d-49cb-bfaa-67c7f61b25a1', '8.10')
Dumping raw-book-json into the bucket at "raw/02776133-d49d-49cb-bfaa-67c7f61b25a1@8.10.json" (application/json)
Requesting baked json book at https://archive.cnx.org/contents/02776133-d49d-49cb-bfaa-67c7f61b25a1@8.10.json
i: ('02776133-d49d-49cb-bfaa-67c7f61b25a1', '8.10')
Dumping baked-book-json into the bucket at "baked/02776133-d49d-49cb-bfaa-67c7f61b25a1@8.10.json" (application/json)
Requesting raw html book at https://archive.cnx.org/contents/02776133-d49d-49cb-bfaa-67c7f61b25a1@8.10.html?as_collated=0
i: ('02776133-d49d-49cb-bfaa-67c7f61b25a1', '8.10')
Dumping raw-book-json into the bucket at "raw/02776133-d49d-49cb-bfaa-67c7f61b25a1@8.10.json" (text/html)
Requesting baked html book at https://archive.cnx.org/contents/02776133-d49d-49cb-bfaa-67c7f61b25a1@8.10.html
i: [('02776133-d49d-49cb-bfaa-67c7f61b25a1', '8.10')]
```

karenc replied:
This conversation is moving to line 246

karenc commented on dump/dump-to-bucket.py:
```diff
@@ -0,0 +1,296 @@
+    # debug(f'Requesting {temperature} HTML {T.bold}{type_}{T.normal} at {T.yellow}{url}{T.normal}')
+    # resp = session.get(url)
+    # yield io.BytesIO(resp.content), 'text/html', f'{temperature}-{type_}-html', ident_hash_seq
+    yield get_content(url, type_, format_, temperature, ident_hash_seq, 'text/html')
```
This is the line that caused the error:

```
Traceback (most recent call last):
  File "./dump-to-bucket.py", line 290, in <module>
    main()
  File "/tmp/rap-spike-lambda/py3/lib/python3.7/site-packages/click/core.py", line 764, in __call__
    return self.main(*args, **kwargs)
  File "/tmp/rap-spike-lambda/py3/lib/python3.7/site-packages/click/core.py", line 717, in main
    rv = self.invoke(ctx)
  File "/tmp/rap-spike-lambda/py3/lib/python3.7/site-packages/click/core.py", line 956, in invoke
    return ctx.invoke(self.callback, **ctx.params)
  File "/tmp/rap-spike-lambda/py3/lib/python3.7/site-packages/click/core.py", line 555, in invoke
    return callback(*args, **kwargs)
  File "./dump-to-bucket.py", line 286, in main
    dump_in_bucket(scrape(book, host), (bucket, region))
  File "./dump-to-bucket.py", line 264, in dump_in_bucket
    key = gen_filepath(type, ident)
  File "./dump-to-bucket.py", line 62, in gen_filepath
    ids = [join_ident_hash(*i) for i in ident]
  File "./dump-to-bucket.py", line 62, in <listcomp>
    ids = [join_ident_hash(*i) for i in ident]
TypeError: join_ident_hash() missing 1 required positional argument: 'version'
```

It needs to set `portion` to `-1` like the other `get_content` calls before this one.

m1yag1 replied:
What line? Could you push it so I don't have to decipher what your talking about?

m1yag1 replied:
n/m figured it out.

m1yag1 replied:
but then i hit another error

```
File "/Users/mike/Projects/00_openstax/00_content_engineering/rap-spike-lambda/dump/dump-to-bucket.py", line 270, in dump_in_bucket
    key = gen_filepath(type, ident)
  File "/Users/mike/Projects/00_openstax/00_content_engineering/rap-spike-lambda/dump/dump-to-bucket.py", line 78, in gen_filepath
    return func(ids)
  File "/Users/mike/Projects/00_openstax/00_content_engineering/rap-spike-lambda/dump/dump-to-bucket.py", line 73, in <lambda>
    'baked-page-json': lambda i: f'baked/{i[0]}:{i[1]}.json',
IndexError: list index out of range
```

mmulich commented:
It looks like I committed the wrong version. So I'll need to back out some `get_content` function changes.

mmulich committed:
Add missing cnx-common dependency

mmulich committed:
Yank out the attemp to refactor the requesting of content

karenc committed:
Add archive host and boto3 aws set up in dump/README

We're using books from staging.cnx.org so we need to set the archive
host to `archive-staging.cnx.org` in the README.  For users who have
never used boto3 before, they need to add their aws credentials to the
boto3 config files.

karenc reviewed (approved):
The script hasn't completed yet for me :joy: but I'm happy with it.

therealmarv reviewed (approved):
now works for me too.
